modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/consistency/GridNearReadRepairAbstractFuture.java

```diff
@@ -228,4 +254,103 @@ else if (!canRemap)
      * Reduces fut's results.
      */
     protected abstract void reduce();
+
+    /**
+     * @param fixedEntries Fixed map.
+     */
+    protected void recordConsistencyViolation(
+        Set<KeyCacheObject> inconsistentKeys,
+        Map<KeyCacheObject, EntryGetResult> fixedEntries
+    ) {
+        GridEventStorageManager evtMgr = ctx.gridEvents();
+
+        if (!evtMgr.isRecordable(EVT_CONSISTENCY_VIOLATION))
+            return;
+
+        Map<Object, Map<ClusterNode, CacheConsistencyViolationEvent.EntryInfo>> originalMap = new HashMap<>();
+
+        for (Map.Entry<ClusterNode, GridPartitionedGetFuture<KeyCacheObject, EntryGetResult>> pair : futs.entrySet()) {
+            ClusterNode node = pair.getKey();
+
+            GridPartitionedGetFuture<KeyCacheObject, EntryGetResult> fut = pair.getValue();
+
+            for (Map.Entry<KeyCacheObject, EntryGetResult> entry : fut.result().entrySet()) {
+                KeyCacheObject key = entry.getKey();
+
+                if (inconsistentKeys.contains(key)) {
+                    EntryGetResult res = entry.getValue();
+                    CacheEntryVersion ver = res.version();
+
+                    Object val = ctx.unwrapBinaryIfNeeded(res.value(), false, false, null);
+
+                    Map<ClusterNode, CacheConsistencyViolationEvent.EntryInfo> map =
+                        originalMap.computeIfAbsent(
+                            ctx.unwrapBinaryIfNeeded(key, false, false, null), k -> new HashMap<>());
+
+                    boolean primary = primaries.get(key).equals(fut.affNode());
```

Comment:
Lets avoid possible NPE
`fut.affNode().equals(primaries.get(key))`
`res.equals(fixedEntries.get(key))`

Reply:
`primaries.get(key)` MUST be not null in this code (because primary should be recorded for each key).
`Objects#equals` overcomplicates the code without real need. 
Let's keep it as is.